### PR TITLE
feat(security): Feature to enable CORS headers for API gateway

### LINKED ISSUE #1913: Enable CORS access to EdgeX microservices
Hello,

# 🚀 Feature Request

### Relevant Package
This feature request is for the API Gateway

### Description
The Edgex components are exposed through REST APIs and those services are usually called from a UI running on a browser.  Most browser prevent calls to services at a different origin (CORS) thus making it difficult the interactions with services like Edgex.
There are basically 2 ways to solve the CORS issue.  The first one is on the client side (UI application) by using a proxy server.  Unfortunately this solution is static and need to know before hand the back end servers.
The second option is to enable the backend server to allow CORS.

### Describe the solution you'd like
Currently I have been bypassing this issue by using a proxy server at the client site.  However, this solution is not working due to the nature of IoT projects and solutions.  The IoT solutions tend to be very dynamic and in my specific example, the number of instances where the Edgex framework is deployed is dynamic and can change.  That makes the proxy option irrelevant as that solution is static.

So, the preferred solution will be to allow the user to enable/disable CORS as desired on the API Gateway.  That will allow calling the API from UI clients without requiring to use a proxy.

### Describe alternatives you've considered
As described above, I am already using a proxy solution but it only works for a static number of Edgex installations.

Thanks in advance,
Marcelo

### PR DESCRIPTION
**Pull request by bnevis-i**
Signed-off-by: Bryon Nevis <bryon.nevis@intel.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/master/.github/Contributing.md.

## What is the current behavior?
Current Kong proxy does not return CORS headers to clients.

## Issue Number:
Closes #1913

## What is the new behavior?
New configuration has been added to allow settings of a limited number of CORS headers for the EdgeX routes.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## New Imports

- [ ] Yes
- [X] No

## Specific Instructions
To test, modify `cmd/security-proxy-setup/res/configuration.toml` and set `EnableCORS` to `true`.  Customize other headers as needed.  Rebuild security-proxy-setup docker container (make docker_security_proxy_setup).  Restart framework.

Test with a command like
```
curl -ki -X OPTIONS -H "Origin: http://localhost" -H "Access-Control-Request-Method: POST" -H "Access-Control-Request-Headers: Content-Type" https://localhost:8443/core-data/api/v2/ping
```

Expected response:
```
HTTP/2 200 
date: Fri, 20 Aug 2021 02:18:02 GMT
vary: Origin
access-control-allow-origin: https://localhost
access-control-allow-headers: Content-Type
access-control-allow-methods: GET,POST,PUT,DELETE
access-control-max-age: 3600
content-length: 0
x-kong-response-latency: 1
server: kong/2.5.0
```

## Other information

### CODE REVIEW

**Comment by bnevis-i** [comment]
@magallardo  Please review to see if this has the functionality that you need.

**Comment by magallardo** [comment]
Great!!   This will address the CORS issue.  What. version should I use for testing?  Hanoi or Geneva?

**Comment by bnevis-i** [comment]
@magallardo  is a feature in Jakarta.  If you need it in Hanoi or Geneva, you will have to cherry-pick it back into your own fork.

File: cmd/security-proxy-setup/res/configuration.toml [hunk 1]
```diff
@@ -38,6 +38,14 @@ Resource = "coredata"
 OutputPath = "accessToken.json"
 JWTFile = "/tmp/edgex/secrets/security-proxy-setup/kong-admin-jwt"
 
+[CORSConfiguration]
+EnableCORS = false
```
Comment (lenny-intel) [inline cmd/security-proxy-setup/res/configuration.toml:42]: Don't we want this enabled by default?

**Comment by lenny-intel** [comment]
> Great!! This will address the CORS issue. What. version should I use for testing? Hanoi or Geneva?

@magallardo , you can test this with the WIP main branch on edgex-compose. You will have to build a local docker image for Proxy Setup from @bnevis-i 's branch and modify the compose file to use that image.
```
make docker_security_proxy_setup
```
Local image will be `edgexfoundry/security-proxy-setup:0.0.0-dev`

File: cmd/security-proxy-setup/res/configuration.toml (hunk 1 above)
Comment (bnevis-i) [reply cmd/security-proxy-setup/res/configuration.toml:42]: I would say no.  Enabling CORS allows additional attack surface and to be effective, the CORS domain list needs to be changed from its default to a customer-specific value.

**Comment by bnevis-i** [comment]
I received a private message that @magallardo will try it out and provide feedback next week.

File: cmd/security-proxy-setup/res/configuration.toml (hunk 1 above)
Comment (lenny-intel) [reply cmd/security-proxy-setup/res/configuration.toml:42]: Need related documentation PR on how to enable CORS

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by magallardo** [comment]
@bnevis-i I have not tried the latest implementation but what we actually did in our deployment and worked fine for us in the Hanoi release is to add the cors plugin to the Kong container.

Following is the command we used:

echo  "Installing cors plugin"
    curl -X POST http://localhost:8001/plugins/ \
    --data "name=cors"  \
    --data "config.origins=http://xxx.xxx.xxx.xxx:xxxx" \
    --data "config.methods=GET" \
    --data "config.methods=POST" \
    --data "config.methods=OPTIONS" \
    --data "config.headers=Accept" \
    --data "config.headers=Accept-Version" \
    --data "config.headers=Content-Length" \
    --data "config.headers=Content-MD5" \
    --data "config.headers=Content-Type" \
    --data "config.headers=Date" \
    --data "config.headers=X-Auth-Token" \
    --data "config.headers=Authorization" \
    --data "config.exposed_headers=X-Auth-Token" \
    --data "config.credentials=true" \
    --data "config.max_age=3600"

 I hope this will help.

I have tried the same command on Ireland version and is not executing and currently trying to find what has changed.

Please let me know if you need anything else.

Thanks,
Marcelo

File: cmd/security-proxy-setup/res/configuration.toml [hunk 2]
```diff
@@ -38,6 +38,14 @@ Resource = "coredata"
...
+CORSAllowCredentials = false
+CORSAllowedOrigins = "https://localhost"
+CORSAllowedMethods = "GET, POST, PUT, DELETE"
```
Comment (cloudxxx8) [inline cmd/security-proxy-setup/res/configuration.toml:45]: We missed `PATCH` here

**Pull request merged**